File: algorithms/leetcode/single_number.py
# Given an array of numbers nums, in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once.

# Example:

# Input:  [1,2,1,3,2,5]
# Output: [3,5]


from collections import Counter
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ans = []
        item_list = Counter(nums)
        
        for key, val in item_list.items():
            if val == 1:
                ans.append(key)
                
        return ans

File: algorithms/leetcode/SingleNumber.cs
/*

Given a non-empty array of integers, every element appears twice except for one. Find that single one.

Note:

Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?

Example 1:

Input: [2,2,1]
Output: 1

Example 2:

Input: [4,1,2,1,2]
Output: 4

*/

// Solution 1 using HashSet
// Runtime O(n)

public class Solution1 {
    
    public int SingleNumber(int[] nums) {
        HashSet<int> myset = new HashSet<int>();
        foreach(int num in nums) {
            if (myset.Contains(num)){
                myset.Remove(num);
            } else {
                myset.Add(num);
            }
        }
        if (myset.Count() == 1) {
            return myset.First();
        }
        return 0;
    }
}

// Solution 2 using Dictionary
// Runtime O(n)
public class Solution2 {
    public int SingleNumber(int[] nums) {
        Dictionary<int, int> myset = new Dictionary<int, int>();
        foreach(int num in nums) {
            if (myset.ContainsKey(num)){
                myset[num]++;
            } else{
                myset.Add(num, 1);            
            }
        }
        foreach(KeyValuePair<int, int> entry in myset) {
            if (entry.Value == 1) {
                return entry.Key;
            }
        }
        return -1;
    }
}